crates/hyperdrive-math/src/long/targeted.rs

```diff
@@ -0,0 +1,518 @@
+use ethers::types::{I256, U256};
+use eyre::{eyre, Result};
+use fixed_point::FixedPoint;
+use fixed_point_macros::fixed;
+
+use crate::{State, YieldSpace};
+
+impl State {
+    /// Gets a target long that can be opened given a budget to achieve a desired fixed rate.
+    pub fn get_targeted_long_with_budget<
+        F1: Into<FixedPoint>,
+        F2: Into<FixedPoint>,
+        F3: Into<FixedPoint>,
+        I: Into<I256>,
+    >(
+        &self,
+        budget: F1,
+        target_rate: F2,
+        checkpoint_exposure: I,
+        maybe_max_iterations: Option<usize>,
+        maybe_allowable_error: Option<F3>,
+    ) -> Result<FixedPoint> {
+        let budget = budget.into();
+        match self.get_targeted_long(
+            target_rate,
+            checkpoint_exposure,
+            maybe_max_iterations,
+            maybe_allowable_error,
+        ) {
+            Ok(long_amount) => Ok(long_amount.min(budget)),
+            Err(error) => Err(error),
+        }
+    }
+
+    /// Gets a target long that can be opened to achieve a desired fixed rate.
+    fn get_targeted_long<F1: Into<FixedPoint>, F2: Into<FixedPoint>, I: Into<I256>>(
+        &self,
+        target_rate: F1,
+        checkpoint_exposure: I,
+        maybe_max_iterations: Option<usize>,
+        maybe_allowable_error: Option<F2>,
+    ) -> Result<FixedPoint> {
+        let target_rate = target_rate.into();
+        let checkpoint_exposure = checkpoint_exposure.into();
+        let allowable_error = match maybe_allowable_error {
+            Some(allowable_error) => allowable_error.into(),
+            None => fixed!(1e14),
+        };
+
+        // Estimate the long that achieves a target rate.
+        let (target_share_reserves, target_bond_reserves) =
+            self.reserves_given_rate_ignoring_exposure(target_rate);
+        let (target_base_delta, target_bond_delta) =
+            self.trade_deltas_from_reserves(target_share_reserves, target_bond_reserves);
+
+        // Determine what rate was achieved.
+        let resulting_rate = self.rate_after_long(target_base_delta, Some(target_bond_delta));
+
+        // The estimated long should always underestimate because the realized price
+        // should always be greater than the spot price.
+        if target_rate > resulting_rate {
+            return Err(eyre!("get_targeted_long: We overshot the zero-crossing.",));
+        }
+        let rate_error = resulting_rate - target_rate;
+
+        // Verify solvency and target rate.
+        if self
+            .solvency_after_long(target_base_delta, target_bond_delta, checkpoint_exposure)
+            .is_some()
```

Comment:
If we are insolvent here and the rate is greater than the desired rate, shouldn't we be using `get_max_long`? Since this isn't using solvency as part of the objective function, it looks like this will be susceptible to failing due to solvency constraints. We should consider having a heuristic that we use to invoke `get_max_long`. Since we're frankly not performance constrained at this point, we could just start with that and see if we have hit the rate yet. I see the merits of optimizing this, so I'll let you make that call.

Reply:
I'd rather this function throws an error if they do not achieve the target instead of giving the max long. If you'd like, I can modify `get_targeted_long_with_budget` to also include a call to `get_max_long` so that it always returns one of the long to hit the target, the budget balance, or the max long.